extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3DataSegmentPusher.java

```diff
@@ -163,21 +155,29 @@ public DataSegment push(final File indexFilesDir, final DataSegment inSegment, f
     );
   }
 
-  private void putObject(String bucketName, String path, S3Object object, boolean replaceExisting)
-      throws ServiceException
+  private void uploadFileIfPossibleAndClear(
+      AmazonS3 s3Client,
+      String bucket,
+      String key,
+      File file,
+      boolean replaceExisting
+  )
   {
-    object.setBucketName(bucketName);
-    object.setKey(path);
-    if (!config.getDisableAcl()) {
-      object.setAcl(GSAccessControlList.REST_CANNED_BUCKET_OWNER_FULL_CONTROL);
-    }
-
-    log.info("Pushing %s.", object);
-
-    if (!replaceExisting && S3Utils.isObjectInBucket(s3Client, bucketName, object.getKey())) {
-      log.info("Skipping push because key [%s] exists && replaceExisting == false", object.getKey());
+    if (!replaceExisting && S3Utils.isObjectInBucketIgnoringPermission(s3Client, bucket, key)) {
+      log.info("Skipping push because key [%s] exists && replaceExisting == false", key);
     } else {
-      s3Client.putObject(bucketName, object);
+      final PutObjectRequest indexFilePutRequest = new PutObjectRequest(bucket, key, file);
+
+      if (!config.getDisableAcl()) {
+        indexFilePutRequest.setAccessControlList(
+            S3Utils.grantFullControlToBucketOwver(s3Client, bucket)
+        );
+      }
+      log.info("Pushing [%s] to bucket[%s] and key[%s].", file, bucket, key);
+      s3Client.putObject(indexFilePutRequest);
     }
+
+    log.info("Deleting temporary cached file [%s]", file.getAbsolutePath());
+    file.delete();
```

Comment:
Would it be better to clean this file up in a finally block?

Reply:
Good point. Fixed.